### PR TITLE
Commit lock files for fuzzers

### PR DESCRIPTION
**Pull request by olix0r**
This change adds fuzz crate lock files to the repo so that
fuzz tests can't regress based on changes in crates.io. It
also removes the per-crate gitignore files by including all
ignored targets in the top-level gitignore.

Finally, we add a CI task to try to build all fuzzers. As this
depends on rust nightly, it will not be a required check.

### CODE REVIEW

**Comment by hawkw** [comment]
> It also removes the per-crate gitignore files by including all
> ignored targets in the top-level gitignore.

FWIW, the per-crate gitignores are automatically generated by `cargo fuzz init` per https://github.com/linkerd/linkerd2-proxy/pull/961#discussion_r605507742, so if we generate more fuzz crates that way, we'll have to remember to remove them. Not a big deal.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -53,6 +53,19 @@ jobs:
...
+    runs-on: ubuntu-latest
+    container:
+      image: docker://rust:1.49.0-buster
+    steps:
+      - uses: actions/checkout@v2
+      - run: rustup toolchain add nightly
+      - run: cargo install cargo-fuzz
+      # Iterate through all fuzz crates to ensure each compiles indpendently.
+      - run: for d in $(find . -name fuzz | sort -r) ; do echo "# $d" ; (cd $d ; cargo +nightly fuzz build) ; done
+      - run: git diff-index --quiet HEAD # Error if the repo isn't clean (i.e. because lockfiles were modified).
```
Comment (hawkw) [inline .github/workflows/rust.yml:67]: could we also use `cargo build --locked here?

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -53,6 +53,19 @@ jobs:
...
+    timeout-minutes: 20
+    runs-on: ubuntu-latest
+    container:
+      image: docker://rust:1.49.0-buster
+    steps:
+      - uses: actions/checkout@v2
+      - run: rustup toolchain add nightly
+      - run: cargo install cargo-fuzz
+      # Iterate through all fuzz crates to ensure each compiles indpendently.
+      - run: for d in $(find . -name fuzz | sort -r) ; do echo "# $d" ; (cd $d ; cargo +nightly fuzz build) ; done
```
Comment (hawkw) [inline .github/workflows/rust.yml:66]: does this need `--features fuzzing` to work?

File: linkerd/dns/Cargo.toml [hunk 3]
```diff
@@ -16,6 +16,6 @@ tokio = { version = "1", features = ["rt", "sync", "time"] }
 pin-project = "1"
 
 [dependencies.trust-dns-resolver]
-version = "0.20.0"
+version = "=0.20.1"
```
Comment (hawkw) [inline linkerd/dns/Cargo.toml:19]: Would be nice to be able to remove this...maybe worth a comment about what upstream changes we are waiting on before we can remove the `=`?

**Review by hawkw — APPROVED** [review]
lgtm, thanks for the lockfile fix! i had a couple comments.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (olix0r) [reply .github/workflows/rust.yml:66]: On closer inspection, we don't actually need this feature at all. I've removed the features and we now rely on `cfg(fuzzing)` gates instead (which also prevents fuzzing deps from being pulled in when running `cargo check`

Comment (hawkw) [reply .github/workflows/rust.yml:66]: Great --- I wasn't entirely sure why there was both a feature flag and a cfg!

File: .github/workflows/rust.yml (hunk 1 above)
Comment (olix0r) [reply .github/workflows/rust.yml]: I'm not sure but I don't think so, due to cfg gates on dependencies.

**Pull request merged**